File: Nanoforge/render/resources/RenderChunk.cpp
#include "RenderChunk.h"
#include "util/Profiler.h"

RenderChunk::RenderChunk(const Mesh& mesh, const RenderChunkData& chunkData, const Vec3& position, const Mat3& rotation)
    : RenderObject(mesh, position, rotation), _chunkData(chunkData)
{

}

void RenderChunk::Draw(ComPtr<ID3D11DeviceContext> d3d11Context, Buffer& perObjectBuffer, Camera& cam)
{
    //TODO: Properly implement for chunks

    if (!Visible)
        return;

    PROFILER_FUNCTION();

    //Bind textures
    if (UseTextures)
    {
        for (u32 i = 0; i < Textures.size(); i++)
        {
            auto texture = Textures[i];
            if (texture.has_value())
                texture.value().Bind(d3d11Context, i);
        }
    }

    for (ChunkPiece& piece : _chunkData.Pieces)
    {
        //Shader constants for this object
        PerObjectConstants constants;

        //Calculate matrices
        DirectX::XMMATRIX objRotation = DirectX::XMMatrixSet
        (
            Rotation.rvec.x, Rotation.rvec.y, Rotation.rvec.z, 0.0f,
            Rotation.uvec.x, Rotation.uvec.y, Rotation.uvec.z, 0.0f,
            Rotation.fvec.x, Rotation.fvec.y, Rotation.fvec.z, 0.0f,
            0.0f, 0.0f, 0.0f, 1.0f
        );
        DirectX::XMMATRIX objTranslation = DirectX::XMMatrixTranslation(Position.x, Position.y, Position.z);
        DirectX::XMMATRIX scale = DirectX::XMMatrixScaling(Scale.x, Scale.y, Scale.z);

        DirectX::XMMATRIX chunkRotation = DirectX::XMMatrixSet
        (
            piece.Orient.rvec.x, piece.Orient.rvec.y, piece.Orient.rvec.z, 0.0f,
            piece.Orient.uvec.x, piece.Orient.uvec.y, piece.Orient.uvec.z, 0.0f,
            piece.Orient.fvec.x, piece.Orient.fvec.y, piece.Orient.fvec.z, 0.0f,
            0.0f, 0.0f, 0.0f, 1.0f
        );
        DirectX::XMMATRIX chunkTranslation = DirectX::XMMatrixTranslation(piece.Position.x, piece.Position.y, piece.Position.z);

        DirectX::XMMATRIX totalRotation = chunkRotation * objRotation;
        DirectX::XMMATRIX totalTranslation = chunkTranslation * objTranslation;

        DirectX::XMMATRIX modelMatrix = DirectX::XMMatrixIdentity();
        modelMatrix = chunkRotation * chunkTranslation * objRotation * objTranslation;


        //Then calculate MVP matrix with Model * View * Projection
        constants.MVP = modelMatrix;//totalRotation * totalTranslation * scale; //First calculate model matrix
        constants.MVP = DirectX::XMMatrixTranspose(constants.MVP * cam.camView * cam.camProjection);
        constants.Rotation = totalRotation;
        constants.WorldPosition.m128_f32[0] = piece.Position.x + Position.x;
        constants.WorldPosition.m128_f32[1] = piece.Position.y + Position.y;
        constants.WorldPosition.m128_f32[2] = piece.Position.z + Position.z;
        constants.WorldPosition.m128_f32[3] = 1.0f;

        //Set constants
        perObjectBuffer.SetData(d3d11Context, &constants);

        //TODO: Get material data for chunks and apply proper textures. Currently it uses the same texture for the entire mesh. 
        //Fill list of submeshes to draw for this chunk piece
        size_t numSubmeshOverrides = 0;
        for (size_t subpieceIndex = piece.FirstSubpiece; subpieceIndex < piece.FirstSubpiece + piece.NumSubpieces; subpieceIndex++)
        {
            u32 submeshIndex = _chunkData.SubpieceSubmeshes[subpieceIndex];
            SubmeshOverrideIndices[numSubmeshOverrides] = submeshIndex;
            numSubmeshOverrides++;
        }

        ObjectMesh.Draw(d3d11Context, numSubmeshOverrides);
    }
}

File: Nanoforge/Render/Resources/RenderChunk.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;
using Nanoforge.Render.Materials;
using Nanoforge.Render.Misc;
using RFGM.Formats.Meshes.Chunks;
using RFGM.Formats.Meshes.Shared;
using Serilog;
using Silk.NET.Vulkan;

namespace Nanoforge.Render.Resources;

public class RenderChunk : RenderObjectBase
{
    public Mesh Mesh;
    public List<Texture2D[]> TexturesByMaterial = [];
    public List<MaterialPipeline> Materials = [];
    
    public Destroyable Destroyable { get; private set; }

    public RenderChunk(Vector3 position, Matrix4x4 orient, Mesh mesh, List<Texture2D[]> texturesByMaterial, string materialName, Destroyable destroyable) : base(position, orient, Vector3.One)
    {
        Position = position;
        Orient = orient;
        Mesh = mesh;
        Destroyable = destroyable;

        const int maxTexturesPerMaterial = 10;
        if (texturesByMaterial.Any(textures => textures.Length > maxTexturesPerMaterial))
        {
            string err = $"Attempted to create a RenderChunk with more than {maxTexturesPerMaterial} textures assigned to a material. They can have {maxTexturesPerMaterial} textures at most.";
            Log.Error(err);
            throw new Exception(err);
        }
        
        foreach (Texture2D[] textures in texturesByMaterial)
        {
            Texture2D[] textureArray = new Texture2D[maxTexturesPerMaterial];
            for (int i = 0; i < textures.Length; i++)
            {
                textureArray[i] = textures[i];
            }
            
            //Use default white texture when one isn't provided. It simplifies the code to just give every RenderObject 10 textures
            if (textures.Length < maxTexturesPerMaterial)
            {
                for (int i = textures.Length; i < maxTexturesPerMaterial; i++)
                {
                    textureArray[i] = Texture2D.MissingTexture;
                }
            }
            
            TexturesByMaterial.Add(textureArray);
        }
        
        foreach (Texture2D[] materialTextures in TexturesByMaterial)
        {
            MaterialPipeline pipeline = MaterialHelper.GetMaterialPipeline(materialName) ?? throw new NullReferenceException($"Failed to get material pipeline: '{materialName}'");
            Materials.Add(pipeline);
        }
        Debug.Assert(Materials.Count > 0);
        Debug.Assert(Materials.Count == TexturesByMaterial.Count);
    }

    public override void WriteDrawCommands(List<RenderCommand> commands, Camera camera, GpuFrameDataWriter constants)
    {
        //Translation and rotation for the overall object
        Matrix4x4 objTranslation = Matrix4x4.CreateTranslation(Position);
        Matrix4x4 objRotation = Orient;
        
        //TODO: Get scaling working
        //Matrix4x4 objScale = Matrix4x4.CreateScale(Scale);
        
        //Setup GPU material data
        List<(int gpuMaterialIndex, MaterialInstance material)> gpuMaterials = new();
        foreach (Texture2D[] textures in TexturesByMaterial)
        {
            MaterialInstance materialInstance = new()
            {
                Texture0 = textures[0].Index,
                Texture1 = textures[1].Index,
                Texture2 = textures[2].Index,
                Texture3 = textures[3].Index,
                Texture4 = textures[4].Index,
                Texture5 = textures[5].Index,
                Texture6 = textures[6].Index,
                Texture7 = textures[7].Index,
                Texture8 = textures[8].Index,
                Texture9 = textures[9].Index,
            };
            int gpuMaterialIndex = constants.AddMaterialInstance(materialInstance);
            gpuMaterials.Add((gpuMaterialIndex, materialInstance));
        }
        
        //Render each chunk piece
        foreach (Dlod dlod in Destroyable.Dlods)
        {
            Matrix4x4 chunkRotation = new Matrix4x4
            (
                dlod.Orient.M11, dlod.Orient.M12, dlod.Orient.M13, 0.0f,
                dlod.Orient.M21, dlod.Orient.M22, dlod.Orient.M23, 0.0f,
                dlod.Orient.M31, dlod.Orient.M32, dlod.Orient.M33, 0.0f,
                0.0f           , 0.0f           , 0.0f           , 1.0f
            );
            
            Matrix4x4 chunkTranslation = Matrix4x4.CreateTranslation(dlod.Pos);
            Matrix4x4 model = chunkRotation * chunkTranslation * objRotation * objTranslation;
            PerObjectConstants baseObjectConstants = new()
            {
                Model = model,
                WorldPosition = new Vector4(Position.X, Position.Y, Position.Z, 1.0f),
            };
            
            for (int subpieceIndex = dlod.FirstPiece; subpieceIndex < dlod.FirstPiece + dlod.MaxPieces; subpieceIndex++)
            {
                ushort submeshIndex = _destroyable.SubpieceData[subpieceIndex].RenderSubpiece;
                SubmeshData submesh = Mesh.Config.Submeshes[submeshIndex];
                uint firstBlock = submesh.RenderBlocksOffset;
                for (int j = 0; j < submesh.NumRenderBlocks; j++)
                {
                    RenderBlock block = Mesh.Config.RenderBlocks[(int)(firstBlock + j)];
                    MaterialPipeline pipeline = Materials[block.MaterialMapIndex];
                    
                    int materialIndex = gpuMaterials[block.MaterialMapIndex].gpuMaterialIndex;
                    PerObjectConstants perObjectConstants = baseObjectConstants with
                    {
                        MaterialIndex = materialIndex,
                    };
                    uint objectIndex = constants.AddObject(perObjectConstants);
                    
                    commands.Add(new RenderCommand
                    {
                        Pipeline = pipeline,
                        Mesh = Mesh,
                        IndexCount = block.NumIndices,
                        StartIndex = block.StartIndex,
                        ObjectIndex = objectIndex,
                    });
                }
            }
        }
    }
    
    public override void Destroy()
    {
        Mesh.Destroy();
        foreach (Texture2D[] textures in TexturesByMaterial)
        {
            foreach (Texture2D texture in textures)
            {
                TextureManager.RemoveReference(texture);
            }
        }
    }
}